### PR TITLE
Threshold SAM3 semantic masks at the model logit threshold

### LINKED ISSUE #25378: SAM3 concept/semantic masks are thresholded on raw logits without sigmoid (logits > 0.5 instead of sigmoid(logits) > 0.5)
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

# SAM3 concept/semantic masks are thresholded on raw logits without `sigmoid` (`logits > 0.5` instead of `sigmoid(logits) > 0.5`)

**Version:** `ultralytics` 8.4.104 · Component: `models/sam` SAM3 semantic path (`SAM3SemanticPredictor`)

### Bug

The SAM3 mask head emits **raw logits** (no activation). In the concept/semantic path they are binarized with a plain `> 0.5` **without `sigmoid`**, so the effective threshold is `sigmoid(logit) > 0.622` instead of the intended `0.5`. Masks come out **systematically eroded** (boundaries pulled inward, thin structures clipped).

The official HF `transformers` reference applies `sigmoid` before thresholding at the same default `mask_threshold=0.5`, and Ultralytics' own **interactive** SAM3 path thresholds correctly (`> mask_threshold`, default `0.0`) — so the two Ultralytics paths are inconsistent.

### Location — `ultralytics/models/sam/predict.py`

```python
# SAM3SemanticPredictor.postprocess (~L2328)
masks = F.interpolate(masks.float()[None], orig_img.shape[:2], mode="bilinear")[0] > 0.5
# SAM3SemanticPredictor.inference_features (~L2403)
pred_masks = F.interpolate(pred_masks.float()[None], src_shape[:2], mode="bilinear")[0] > 0.5
```

Official (`transformers/models/sam3/image_processing_sam3.py`): `pred_masks.sigmoid()` then `masks > mask_threshold` (default `0.5`).

### Fix

Add `.sigmoid()` before `> 0.5` in both locations (or threshold raw logits at `> 0.0`):

```diff
- ...mode="bilinear")[0] > 0.5
+ ...mode="bilinear")[0].sigmoid() > 0.5
```

### Environment
  - Runtime: Docker container
  - OS: Ubuntu 22.04.3 LTS (Jammy), x86_64
  - Kernel: Linux 4.15.0-213-generic
  - Python: 3.10.12
    - Executable: `/usr/bin/python3.10`
    - Compiler: GCC 11.4.0
  - pip: 26.1.2
    - Location: `/usr/local/lib/python3.10/dist-packages/pip`
  - Installation method: pip
  - Ultralytics: 8.4.104
  - PyTorch: 2.5.1+cu121
  - torchvision: 0.20.1+cu121
  - NumPy: 2.2.6
  - OpenCV: 5.0.0.93
  - Pillow: 12.2.0
  - CUDA available: Yes
  - PyTorch CUDA build: 12.1
  - NVIDIA driver: 535.146.02
  - Maximum CUDA version reported by `nvidia-smi`: 12.2
  - cuDNN: 9.1.0
  - CUDA toolkit / `nvcc`: Not installed or not available in PATH
  - GPUs: 4 × NVIDIA GeForce RTX 3090, 24 GB each
  - CPU: AMD Ryzen Threadripper PRO 3995WX
    - 64 physical cores / 128 logical CPUs
  - RAM: 125.8 GB
  - Swap: None
  - SAM3 input size: 1008
  - SAM3 weight: `sam3.pt`

  `ultralytics.checks()` output:

  ```text
  Ultralytics 8.4.104 🚀 Python-3.10.12 torch-2.5.1+cu121
  CUDA:0 (NVIDIA GeForce RTX 3090, 24260MiB)
  Setup complete ✅ (128 CPUs, 125.8 GB RAM)
  ```

### Minimal Reproducible Example
Reproduction (no weights needed)
```python
import torch
logits = torch.linspace(-3, 3, steps=13)      # raw mask-head logits
ultra    = logits > 0.5                        # Ultralytics semantic path (bug)
official = logits.sigmoid() > 0.5              # official  (== logits > 0.0)
print("disagree:", (ultra != official).int().tolist())
# pixels with logit in (0, 0.5] -> official=1, ultralytics=0 -> eroded
```
output
```python
logit          : [-3.0, -2.5, -2.0, -1.5, -1.0, -0.5, 0.0, 0.5, 1.0, 1.5, 2.0, 2.5, 3.0]
sigmoid(logit) : [0.047, 0.076, 0.119, 0.182, 0.269, 0.378, 0.5, 0.622, 0.731, 0.818, 0.881, 0.924, 0.953]
ultralytics    : [0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1]
official       : [0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1]
disagree       : [0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0]
```

### Additional

_No response_

### Are you willing to submit a PR?

- [o] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Y-T-G**
Fixes #25378.

## Summary

`SAM3SemanticPredictor` was thresholding raw mask logits at `0.5`, which is an effective probability cutoff of about `0.622` and erodes mask boundaries. Both semantic output paths now use the model's existing `mask_threshold` (`0.0`), matching the other SAM owners and remaining equivalent to `sigmoid(logits) > 0.5` without adding a sigmoid kernel.

Deleted: the two incorrect hard-coded `0.5` logit thresholds; no helper, branch, or test code was added.

## Validation

- `ruff check ultralytics/models/sam/predict.py`
- Verified `logits > 0.0` is exactly equal to `logits.sigmoid() > 0.5` across representative logits.
- Full CI is green on the exact head.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
🛠️ SAM mask postprocessing now uses the model’s configurable `mask_threshold` instead of a fixed `0.5` threshold.

### 📊 Key Changes
- Replaced hardcoded mask binarization thresholds in both standard prediction and feature-based inference.
- Applied `self.model.mask_threshold` when converting interpolated mask probabilities into binary masks.
- Kept existing mask resizing and bounding-box scaling behavior unchanged.

### 🎯 Purpose & Impact
- 🎯 Allows SAM models to control mask sensitivity through a configurable threshold.
- ✅ Improves consistency between model configuration and inference results.
- 🔧 Enables users to adjust segmentation precision and recall without modifying source code.
- 🚀 Affects generated masks directly, potentially changing object boundaries and the number of pixels classified as foreground.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution has been received, and an engineer will assist with the review. Please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The fix correctly identifies that SAM3 masks are logits, but applying sigmoid solely before a 0.5 boolean threshold is redundant. Use a zero logit threshold in both paths to retain the same mask boundary with less inference overhead.

💬 Posted 1 inline comment

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Clean, focused fix. Both SAM3 semantic-mask output paths now use the model’s established logit threshold, consistent with the SAM mask-threshold convention; no verified issues found.

**Comment by UltralyticsAssistant** [comment]
🎉 Thank you @Y-T-G for this valuable contribution, now merged!

> “Great things are done by a series of small things brought together.” — Vincent van Gogh

Using the configurable `mask_threshold` throughout SAM mask postprocessing is a focused improvement with meaningful impact, giving users consistent control over segmentation sensitivity while preserving existing resizing and bounding-box behavior. Your attention to this detail helps make Ultralytics models more flexible and reliable. 🚀
